### ISSUE TITLE
Request: add file browser file previews to the Nextcloud whitelist

### ISSUE THREAD

**Issue by bdcarr** [crowdsecurity/hub#627]
The [Nextcloud whitelist collection](https://github.com/crowdsecurity/hub/pull/609) was created to work around an issue in the [Contacts](https://github.com/nextcloud/contacts/issues/3021) and [Photos](https://github.com/nextcloud/photos/issues/1257) apps in Nextcloud. I've just learned that the same issue occurs in the Files app - probably anywhere a file preview is loaded really.

I just banned myself while searching for files by clicking 'load more' a couple of times with XML files in the search results:

![image](https://user-images.githubusercontent.com/1940488/210563018-d6d82e41-795c-4f9f-83bb-7c2b6b29a864.png)

An example request URL is `https://nextcloud.mydomain.com/core/preview?x=32&y=32&fileId=60489`. I might have a crack at updating the whitelist parser & tests sometime in the next few days, but I'm logging this issue in case I forget or don't have time.

**Comment by LaurenceJJones** [comment]
I added some extra stuff since you linked PR, I appended a potential fix but some logs / tests will be useful before committing a PR
```
name: crowdsecurity/nextcloud-whitelist
description: "Whitelist events from nextcloud"
filter: "evt.Meta.service == 'http' && evt.Meta.log_type in ['http_access-log', 'http_error-log']"
whitelist:
  reason: "Nextcloud Whitelist"
  expression:
   - evt.Meta.http_status == '404' && evt.Meta.http_verb == 'GET' && evt.Parsed.file_ext == '.vcf' && evt.Parsed.http_args contains "photo" #Contacts app .vcf missing photo
   - evt.Meta.http_status == '404' && evt.Meta.http_verb == 'GET' && evt.Meta.http_path contains '/apps/files_versions/preview' && evt.Parsed.http_args contains 'version' #Backup app missing file version
   - evt.Meta.http_status == '404' && evt.Meta.http_verb == 'GET' && evt.Meta.http_path contains '/apps/photos/api/v1/preview' && evt.Parsed.http_args contains 'x' && evt.Parsed.http_args contains 'y' #Photo app loads all previews as small panes, but can 404
   - evt.Meta.http_status == '404' && evt.Meta.http_verb == 'GET' && evt.Meta.http_path contains '/core/preview' && evt.Parsed.http_args contains 'x' && evt.Parsed.http_args contains 'y' #File preview, but can 404
```

**Comment by bdcarr** [comment]
That's wonderful @LaurenceJJones, thanks for acting so quickly. I haven't had the bandwidth to learn how parsers etc work yet, sorry. Just popping in to note that we should probably exclude PROPFIND requests to `/remote.php/webdav/*` that get 404 responses (but not 403). I just tried to upload a bunch of files to my nextcloud server and got banned :disappointed:

**Comment by bdcarr** [comment]
I started on adding new expressions to the whitelist. Having some issues with running tests so I don't know if these work yet. Posted in discord about it.

```
   - evt.Meta.http_status == '404' && evt.Meta.http_verb == 'GET' && evt.Meta.http_path == '/core/preview' && evt.Parsed.http_args == ['x', 'y', 'fileId'] #File preview often 404s while searching
   - evt.Meta.http_status == '404' && evt.Meta.http_verb == 'PROPFIND' && evt.Meta.http_path startsWith '/remote.php/webdav/' #Uploading new files via WebDAV always produces a 404
```

I'm trying to make the exclusions as narrow as possible, so I replaced the "contains" checks with an array equality check for the http_args since as far as I can see, the file preview request always has those 3. Extra/missing args could be actual HTTP probing.

**Comment by LaurenceJJones** [comment]
>I'm trying to make the exclusions as narrow as possible, so I replaced the "contains" checks with an array equality check for the http_args since as far as I can see

This wont work as http_args is a pipe seperaated list for example

```
http://example.com/?a=foo&b=bar -> "a=foo|b=bar"
```

That is why my original just said contains as its a string.

**Comment by bdcarr** [comment]
[The expressions seem good now](https://github.com/bdcarr/crowdsec-hub/blob/09b3d9dd8852714e30ff1b25330e0e9f1645a0c0/parsers/s02-enrich/crowdsecurity/nextcloud-whitelist.yaml). Just waiting on [a reply in discord](https://discord.com/channels/921520481163673640/1069062913203118100/1069062913203118100) about testing.

**Comment by chadsell** [comment]
I am still getting banned due to this preview 404 issue.  I confirmed that I have the latest version of this whitelist. The only thing I can think of that might be my issue is that I am running the "NextCloudPi" version of nextcloud, and my http logs are in a file called nc-access.log instead of the regular access.log.  Both *access.log files are in the same /var/log/apache2/ directory. Is it possible that the whitelist is not applying to the nc-access.log file? How can I fix this?

**Comment by LaurenceJJones** [comment]
> I am still getting banned due to this preview 404 issue. I confirmed that I have the latest version of this whitelist. The only thing I can think of that might be my issue is that I am running the "NextCloudPi" version of nextcloud, and my http logs are in a file called nc-access.log instead of the regular access.log. Both *access.log files are in the same /var/log/apache2/ directory. Is it possible that the whitelist is not applying to the nc-access.log file? How can I fix this?

Open a new issue and provide details behind the alert `cscli alerts list` (find the alert ID) then run `cscli alerts inspect <id> -d` you can omit the IP address.
